[BUGFIX] Preserve all uninlinable CSS at-rules

JakeQZ opened the pull request:
All at-rules that have no special handling by Emogrifier will now by copied to a
`<style>` element.  Previously, Emogrifier would attempt to process them as
regular CSS rules, causing an exception in `CssSelectorConverter::toXPath()`,
which in 'debug' mode would be propagated; in either case they would be
discarded.

The rationale is that if any such rules are in the supplied CSS, they are
desired in the resulting HTML.  Emogrifier can treat them as a black box it
knows nothing about, simply to be passed on.

For at-rules which are containers for regular CSS rules (e.g. `@supports`), the
ordering could be important.  For now, these rules will be placed at the start
of the `<style>` element, so it is possible that they will be overridden by
other uninlinable rules that are placed after them, but were originally placed
before.  However, there should be no regression because, by being placed first
(and in their own original order), they would still never override a rule in the
`<style>` element that they would not have in the original CSS.

The ordering improvement can be addressed separately, e.g. by handling container
at-rules, like `@supports`, more generally, in a similarly way to how `@media`
rules are currently processed.  (Addressing #544 first may also help.)

Fixes #886.

oliverklee commented on src/CssInliner.php:
```diff
@@ -523,38 +523,44 @@ private function extractImportAndCharsetRules(string $css): array
         while (
             \preg_match(
-                '/(@font-face[^}]++}\\s*+)/i',
+                '/(@(?!media\\b)[\\w\\-]++)[^\\{]*+(\\{[^\\{\\}]*+(?:(?2)[^\\{\\}]*+)*+\\})\\s*+/i',
```
Please let's move this to a private const so that the method has only the code, but no data.

JakeQZ replied:
Done.  This also allows the preceding comment to be moved to the DocBlock for the `const`.

oliverklee commented on src/CssInliner.php:
```diff
@@ -523,38 +523,44 @@ private function extractImportAndCharsetRules(string $css): array
-                $fontFaces .= $atRuleAndFollowingWhitespace;
+            if (
+                \strcasecmp($atRuleName, '@font-face') !== 0
+                || \stripos($fullMatch, 'font-family') !== false && \stripos($fullMatch, 'src') !== false
```
Can we move this to a small, well-named method?

JakeQZ replied:
Done.

oliverklee commented on tests/Unit/CssInlinerTest.php:
```diff
@@ -52,6 +52,100 @@ final class CssInlinerTest extends TestCase
+    /**
+     * @var string[]
+     *
+     * selectoin of at-rules which have no special handling by `CssInliner` but should be passed through and placed in a
```
typo: selectoin -> selection

JakeQZ replied:
Fixed.  Well spotted.

JakeQZ commented:
> 
> 
> Looks mostly very good. I've added some comments. This PR also needs a rebase.

I've made changes to address the comments and rebased.

I forgot to use the `-force` when rebasing then followed the incorrect suggestion to execute `git pull` - [déjà vu](https://github.com/MyIntervals/emogrifier/issues/373#issuecomment-536631092) - but all should hopefully be sorted now.
